File: Braintree/src/main/java/com/braintreepayments/api/models/VisaCheckoutConfiguration.java
package com.braintreepayments.api.models;

import com.braintreepayments.api.Json;
import com.braintreepayments.api.internal.ClassHelper;
import com.braintreepayments.api.internal.VisaCheckoutConstants;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Contains the remote Visa Checkout configuration for the Braintree SDK.
 */
public class VisaCheckoutConfiguration {

    private boolean mIsEnabled;
    private String mApiKey;
    private String mExternalClientId;
    private List<String> mCardBrands;

    static VisaCheckoutConfiguration fromJson(JSONObject json) {
        VisaCheckoutConfiguration visaCheckoutConfiguration = new VisaCheckoutConfiguration();

        if (json == null) {
            json = new JSONObject();
        }

        visaCheckoutConfiguration.mApiKey = Json.optString(json, "apikey", "");
        visaCheckoutConfiguration.mIsEnabled = ClassHelper.isClassAvailable(VisaCheckoutConstants.VISA_CHECKOUT_CLASSNAME) &&
                !visaCheckoutConfiguration.mApiKey.equals("");
        visaCheckoutConfiguration.mExternalClientId = Json.optString(json, "externalClientId", "");
        visaCheckoutConfiguration.mCardBrands = supportedCardTypesToAcceptedCardBrands(
                CardConfiguration.fromJson(json).getSupportedCardTypes());

        return visaCheckoutConfiguration;
    }

    /**
     * Determines if the Visa Checkout flow is available to be used. This can be used to determine
     * if UI components should be shown or hidden.
     *
     * @return boolean if Visa Checkout SDK is available, and configuration is enabled.
     */
    public boolean isEnabled() {
        return mIsEnabled;
    }

    /**
     * @return The Visa Checkout External Client Id associated with this merchant's Visa Checkout configuration.
     */
    public String getExternalClientId() {
        return mExternalClientId;
    }

    /**
     * @return The Visa Checkout API Key associated with this merchant's Visa Checkout configuration.
     */
    public String getApiKey() {
        return mApiKey;
    }

    /**
     * @return The accepted card brands for Visa Checkout.
     */
    public List<String> getAcceptedCardBrands() {
        return mCardBrands;
    }

    private static List<String> supportedCardTypesToAcceptedCardBrands(Set<String> supportedCardTypes) {
        List<String> acceptedCardBrands = new ArrayList<>();

        for (String supportedCardType : supportedCardTypes) {
            switch (supportedCardType.toLowerCase(Locale.ROOT)) {
                case "visa":
                    acceptedCardBrands.add("VISA");
                    break;
                case "mastercard":
                    acceptedCardBrands.add("MASTERCARD");
                    break;
                case "discover":
                    acceptedCardBrands.add("DISCOVER");
                    break;
                case "american express":
                    acceptedCardBrands.add("AMEX");
                    break;
            }
        }

        return acceptedCardBrands;
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/VisaCheckoutConfiguration.kt
package com.braintreepayments.api

import org.json.JSONObject
import java.util.ArrayList

/**
 * Contains the remote Visa Checkout configuration for the Braintree SDK.
 * @property isEnabled Determines if the Visa Checkout flow is available to be used. This can be used to determine
 *                      if UI components should be shown or hidden.
 * @property apiKey The Visa Checkout API Key associated with this merchant's Visa Checkout configuration.
 * @property externalClientId The Visa Checkout API Key associated with this merchant's Visa Checkout configuration.
 * @property acceptedCardBrands The accepted card brands for Visa Checkout.
 */
internal data class VisaCheckoutConfiguration(
    val apiKey: String,
    val externalClientId: String,
    val acceptedCardBrands: List<String>
) {

    val isEnabled: Boolean = apiKey != ""

    constructor(json: JSONObject?) : this(
        Json.optString(json, API_KEY, ""),
        Json.optString(json, EXTERNAL_CLIENT_ID, ""),
        supportedCardTypesToAcceptedCardBrands(
            CardConfiguration(json).supportedCardTypes
        )
    )

    companion object {

        private const val API_KEY = "apikey"
        private const val EXTERNAL_CLIENT_ID = "externalClientId"

        private fun supportedCardTypesToAcceptedCardBrands(supportedCardTypes: List<String>): List<String> {
            val acceptedCardBrands: MutableList<String> = ArrayList()
            for (supportedCardType in supportedCardTypes) {
                when (supportedCardType.lowercase()) {
                    "visa" -> acceptedCardBrands.add("VISA")
                    "mastercard" -> acceptedCardBrands.add("MASTERCARD")
                    "discover" -> acceptedCardBrands.add("DISCOVER")
                    "american express" -> acceptedCardBrands.add("AMEX")
                }
            }
            return acceptedCardBrands
        }
    }
}
